core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -125,43 +133,46 @@ public void setStoreListenerSource(Cache<K, V> source) {
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    eventThreadLocal.get().add(cacheEventWrapper);
+    if (!orderedListenersSet.isEmpty()) {
+      ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue;
+      if(keyBasedEventQueueMap.keySet().contains(event.getKey())) {
+        eventWrapperQueue = keyBasedEventQueueMap.get(event.getKey());
+      } else {
+        eventWrapperQueue = new ConcurrentLinkedQueue<CacheEventWrapper>();
+        keyBasedEventQueueMap.put(event.getKey(), eventWrapperQueue);
+      }
+      eventWrapperQueue.add(cacheEventWrapper);
+    }
+  }
+  
+  private void dispatchEvent(CacheEventWrapper eventWrapper) {
+    ConcurrentLinkedQueue<CacheEventWrapper> eventWrapperQueue = keyBasedEventQueueMap.get(eventWrapper.cacheEvent.getKey());
+    boolean interrupted = false;
+    for (EventListenerWrapper wrapper: registeredListeners) {
+      if (!wrapper.config.fireOn().contains(eventWrapper.cacheEvent.getType())) {
         continue;
       }
-      Runnable notificationTask = new Runnable() {
-        @Override
-        public void run() {
-          CacheEventListener<K, V> listener = wrapper.getListener();
-          listener.onEvent(event);
-        }
-      };
-      
-      ExecutorService eventDelivery = wrapper.config.orderingMode().equals(EventOrdering.UNORDERED) ? unorderedDelivery : orderedDelivery;
-      notificationResults.put(wrapper, eventDelivery.submit(notificationTask));
-    }
-    
-    for (Map.Entry<EventListenerWrapper, Future<?>> entry: notificationResults.entrySet()) {
-      EventListenerWrapper wrapper = entry.getKey();
-      Future<?> f = entry.getValue();
-      if ((EventFiring.SYNCHRONOUS.equals(wrapper.config.firingMode()))) {
-        boolean interrupted = false;
+      if( wrapper.config.firingMode() == EventFiring.SYNCHRONOUS) {
         try {
-          f.get();
-        } catch (ExecutionException e) {
-            LOGGER.error("Cache Event Listener failed for event type {} due to ", type, e);
-          // XXX delegate to resilience strategy (#52), and/or just log
-        } catch (InterruptedException e) {
-          interrupted = true;
-        } finally {
-          if (interrupted) {
-            Thread.currentThread().interrupt();
+          Future future = eventDispatcher.dispatch(eventWrapper, wrapper.getListener());
+          future.get();
+        } catch (InterruptedException ie) {
+          if(interrupted) {
+            break;
           }
-        }
+          interrupted = true;
+        } catch (ExecutionException ee) {
+          LOGGER.error("Cache Event Listener failed for event type {} due to ", eventWrapper.cacheEvent.getType(), ee);
+        } 
+      } else {
+        eventDispatcher.dispatch(eventWrapper, wrapper.getListener());
       }
+      
+    }
+    if (eventWrapperQueue != null && !eventWrapperQueue.isEmpty() && eventWrapperQueue.peek() == eventWrapper.cacheEvent) {
```

Comment:
How do you respect per key ordering across threads if you do not even check this var before now?